WebKit should provide a move option for setting custom user agent (and other things)

brentfulgham opened the pull request:
#### 84e85f84be9c1913afc0222aae54a7b0fad53c48
<pre>
WebKit should provide a move option for setting custom user agent (and other things)
<a href="https://bugs.webkit.org/show_bug.cgi?id=291285">https://bugs.webkit.org/show_bug.cgi?id=291285</a>
<a href="https://rdar.apple.com/148857394">rdar://148857394</a>

Reviewed by Geoffrey Garen.

While working on Bug 291113 I realized that DocumentLoader and RemoteFrame do not
provide a move overload, even though they are often populated with the result of a
new temporary string. We could avoid copying a non-trivial string in a few places
by correcting this.

I also noticed a few similar patterns and added Move operators for them.

* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::setResponse):
(WebCore::DocumentLoader::clientRedirectSourceForHistory const):
(WebCore::DocumentLoader::setClientRedirectSourceForHistory):
(WebCore::DocumentLoader::serverRedirectDestinationForHistory const):
(WebCore::DocumentLoader::setCustomUserAgent):
* Source/WebCore/loader/FrameLoadRequest.cpp:
(WebCore::FrameLoadRequest::FrameLoadRequest):
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::setClientRedirectSourceForHistory):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::reloadWithOverrideEncoding):
(WebCore::FrameLoader::reload):
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
(WebCore::FrameLoader::loadDifferentDocumentItem):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::initializeItem):
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::didReceiveResponse):
* Source/WebCore/loader/appcache/ApplicationCacheHost.cpp:
(WebCore::ApplicationCacheHost::maybeLoadMainResource):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::load):
(WebCore::CachedResource::setResponse):
(WebCore::CachedResource::setResponseInternal):
(WebCore::CachedResource::responseReceived):
* Source/WebCore/loader/cache/CachedResource.h:
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::CachedResourceRequest::splitFragmentIdentifierFromRequestURL):
(WebCore::upgradeInsecureResourceRequestIfNeeded):
(WebCore::CachedResourceRequest::removeFragmentIdentifierIfNeeded):
* Source/WebCore/loader/cocoa/BundleResourceLoader.mm:
(WebCore::BundleResourceLoader::loadResourceFromBundle):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::upgradeInsecureRequestIfNeeded const):
* Source/WebCore/platform/ContextMenuItem.h:
(WebCore::ContextMenuItem::setTitle):
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
(WebCore::NetworkExtensionContentFilter::willSendRequest):
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::setURL):
* Source/WebCore/platform/network/ResourceRequestBase.h:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::ResourceResponseBase):
(WebCore::m_isNull):
* Source/WebCore/platform/network/ResourceResponseBase.h:
* Source/WebCore/platform/network/cf/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse):
* Source/WebCore/platform/network/mac/ResourceHandleMac.mm:
(WebCore::ResourceHandle::createNSURLConnection):
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::applyToDocumentLoader):
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCustomUserAgent):
(WebKit::WebPageProxy::setCustomUserAgentInternal):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::applyWebsitePolicies):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):
(WebKit::WebPage::loadData):
(WebKit::WebPage::setUserAgent):
(WebKit::WebPage::setUserAgentInternal):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/294915@main">https://commits.webkit.org/294915@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03c5bd03c20a9fea250265ba8464c7e9fc51a076

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103508 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23190 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13510 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108684 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54153 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105547 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23542 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31741 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106514 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/18259 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93381 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/58994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11428 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53509 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/96183 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11487 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111062 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102119 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30655 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31016 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89581 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/87299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22221 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32173 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/9890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25006 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30583 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/35895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125752 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30383 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34841 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33714 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31944 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4428,12 +4428,26 @@ String Document::documentURI() const
     );
 }
 
+void Document::setURL(URL&& url)
```
These PRs are causing a lot of code duplication without performance numbers. This is very unfortunate. Several ways to avoid that:
1. Keep only the move version and port all call sites to it
2. Update the parameter to pass the URL by value. Then the internal implementation would do a WTFMove(). And the call sites can either WTFMove() the parameter in or not, with a single implementation.

cc @geoffreygaren @darinadler @weinig for opinions.

darinadler replied:
I strongly agree with your overall point, @cdumez. I also agree that either of your suggestions would be a good approach. Your suggestion (2) is more elegant at call sites, and your suggestion (1) is less elegant but slightly decreases the chance that we’ll accidentally keep reference count churn in cases where we can avoid it. I think I prefer (2).

geoffreygaren replied:
Yeah, in cases with mixed callers (T& and T&&) I like (2). You can always switch to (1) eventually if all callers become T&&.

brentfulgham replied:
@cdumez So if I modify this code so that instead of a (T&&) and a (const T&) implementation, we just have a single (T) implementation, moves will happen in all cases where we remember to WTFMove() the argument, or it the result of a RVO call?

brentfulgham replied:
@darinadler I now wonder if there is ever a reason to use a (const T&) setter rather than (T) with a move in the implementation, since setters always pay the price of copying into the local object so there's no real benefit of the reference being passed into the argument. Unless the setter can use the passed argument to make a decision that reduces the need for the copy.

mcatanzaro replied:
> So if I modify this code so that instead of a (T&&) and a (const T&) implementation, we just have a single (T) implementation, moves will happen in all cases where we remember to WTFMove() the argument, or it the result of a RVO call?

Yes.

Notably, pass by value is only a good approach here because you are always going to either (a) move the URL, or (b) copy the URL. Normally you would only want to use pass by value with primitive types or very small objects, but `Document::setURL` is going to copy that URL one way or the other, so it really makes sense to just use pass by value here.

cdumez replied:
> moves will happen in all cases where we remember to WTFMove() the argument

Right. There will be zero copy if the call sites uses WTFMove(). There will be a single copy when they don't.

mcatanzaro replied:
Note there are [core guidelines](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#fcall-parameter-passing) covering this. I ought to read those....

darinadler replied:
A more difficult case than this one, which presents a small design challenge, is a function that often examines its argument and only sometimes moves the value of argument in. In that case a `const T&` can be more efficient than `T` for the case where the argument value is not moved in, so it's trickier to decide which solution is best.

geoffreygaren replied:
Another tricky case is when T is a large type (like FontDescription) rather than a pointer wrapper (like URL). In that case, I’m not sure that a WTFMove inside the callee is enough to avoid copying memory.

darinadler replied:
> Another tricky case is when T is a large type (like FontDescription) rather than a pointer wrapper (like URL). In that case, I’m not sure that a WTFMove inside the callee is enough to avoid copying memory.

Excellent point. Here’s the shorthand way I think about that: If it’s not inlined or link-time optimized there are 2 moves and 0 copies in this idiom. Thus, if the move is expensive enough, might need to use `T&&` so it’s only 1 move and 0 copies.

mcatanzaro replied:
URL is actually not just a pointer wrapper: [it's filled with integers](https://github.com/WebKit/WebKit/blob/fa3ef110e7667e0299e44ed2959ac1e4890f4154/Source/WTF/wtf/URL.h#L272).

cdumez replied:
> URL is actually not just a pointer wrapper: [it's filled with integers](https://github.com/WebKit/WebKit/blob/fa3ef110e7667e0299e44ed2959ac1e4890f4154/Source/WTF/wtf/URL.h#L272).

True, URL is actually quite large. The good news is that I believe Brent is working on updating his PR to use `T&&` only.

geoffreygaren commented on Source/WTF/wtf/LogChannels.cpp:
```diff
@@ -61,7 +61,7 @@ void LogChannels::clearAllLogChannelsToAccumulate()
 }
 
-void LogChannels::initializeLogChannelsIfNecessary(std::optional<String> logChannelString)
+void LogChannels::initializeLogChannelsIfNecessary(const std::optional<String>& logChannelString)
```
Maybe this should be `const String&`? `String` is nullable (`String().isNull() == true`). An `optional` to a nullable type is a nullable nullable type, which hurts my brain. Technically the code below is incomplete, since it only checks for null, but not non-null + null. The brain pain increases.

brentfulgham replied:
No objections from me to make that change.

geoffreygaren commented on Source/WebKit/Shared/SessionState.h:
```diff
@@ -125,12 +125,17 @@ class FrameState : public RefCounted<FrameState> {
     Vector<Ref<FrameState>> children;
 
 private:
+    FrameState(String&& urlString, String&& originalURLString, String&& referrer, AtomString&& target, std::optional<WebCore::FrameIdentifier>, std::optional<Vector<uint8_t>>&& stateObjectData, int64_t documentSequenceNumber, int64_t itemSequenceNumber, WebCore::IntPoint scrollPosition, bool shouldRestoreScrollPosition, float pageScaleFactor, std::optional<HTTPBody>&&, std::optional<WebCore::Back ...
```
Does this get called somewhere in this patch? Cmd-F isn't finding it. (If not called, probably best to leave out.)

brentfulgham replied:
It gets used by the GeneratedSerializer IPC code, which initializes a FrameState using a bunch of Moves of the decoded members. I put a breakpoint in the old constructor (before adding this) and hit it in the serializer, so it seemed like we were doing a bunch of copies we don't need to.

brentfulgham committed:
WebKit should provide a move option for setting custom user agent (and other things)
https://bugs.webkit.org/show_bug.cgi?id=291285
rdar://148857394

Reviewed by Geoffrey Garen.

While working on Bug 291113 I realized that DocumentLoader and RemoteFrame do not
provide a move overload, even though they are often populated with the result of a
new temporary string. We could avoid copying a non-trivial string in a few places
by correcting this.

I also noticed a few similar patterns and added Move operators for them.

* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::setResponse):
(WebCore::DocumentLoader::clientRedirectSourceForHistory const):
(WebCore::DocumentLoader::setClientRedirectSourceForHistory):
(WebCore::DocumentLoader::serverRedirectDestinationForHistory const):
(WebCore::DocumentLoader::setCustomUserAgent):
* Source/WebCore/loader/FrameLoadRequest.cpp:
(WebCore::FrameLoadRequest::FrameLoadRequest):
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::setClientRedirectSourceForHistory):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::reloadWithOverrideEncoding):
(WebCore::FrameLoader::reload):
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
(WebCore::FrameLoader::loadDifferentDocumentItem):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::initializeItem):
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::SubresourceLoader::didReceiveResponse):
* Source/WebCore/loader/appcache/ApplicationCacheHost.cpp:
(WebCore::ApplicationCacheHost::maybeLoadMainResource):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::load):
(WebCore::CachedResource::setResponse):
(WebCore::CachedResource::setResponseInternal):
(WebCore::CachedResource::responseReceived):
* Source/WebCore/loader/cache/CachedResource.h:
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::CachedResourceRequest::splitFragmentIdentifierFromRequestURL):
(WebCore::upgradeInsecureResourceRequestIfNeeded):
(WebCore::CachedResourceRequest::removeFragmentIdentifierIfNeeded):
* Source/WebCore/loader/cocoa/BundleResourceLoader.mm:
(WebCore::BundleResourceLoader::loadResourceFromBundle):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::upgradeInsecureRequestIfNeeded const):
* Source/WebCore/platform/ContextMenuItem.h:
(WebCore::ContextMenuItem::setTitle):
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
(WebCore::NetworkExtensionContentFilter::willSendRequest):
* Source/WebCore/platform/network/ResourceRequestBase.cpp:
(WebCore::ResourceRequestBase::setURL):
* Source/WebCore/platform/network/ResourceRequestBase.h:
* Source/WebCore/platform/network/ResourceResponseBase.cpp:
(WebCore::ResourceResponseBase::ResourceResponseBase):
(WebCore::m_isNull):
* Source/WebCore/platform/network/ResourceResponseBase.h:
* Source/WebCore/platform/network/cf/ResourceResponse.h:
(WebCore::ResourceResponse::ResourceResponse):
* Source/WebCore/platform/network/mac/ResourceHandleMac.mm:
(WebCore::ResourceHandle::createNSURLConnection):
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::applyToDocumentLoader):
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCustomUserAgent):
(WebKit::WebPageProxy::setCustomUserAgentInternal):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::applyWebsitePolicies):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::loadRequest):
(WebKit::WebPage::loadData):
(WebKit::WebPage::setUserAgent):
(WebKit::WebPage::setUserAgentInternal):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/294915@main
